### ISSUE TITLE
Save and get the selected language just as the theme switch 

### ISSUE THREAD

**Issue by d-oit** [gethinode/hinode#1024]
## Problem or enhancement idea

The current version of hinode doesn't save the selected language like the theme switch.

## Proposed solution

Save and load the selected language just as the theme switch in the localStorage

### Discussed in https://github.com/gethinode/hinode/discussions/1023

<div type='discussions-op-text'>

<sup>Originally posted by **d-oit** July  4, 2024</sup>
Hello Mark,

first of all great theme. Appreciate your work.

Is there a reason why the selected language is not stored in the localStorage.

Thanks for your reply.

Dominik </div>

**Comment by d-oit** [comment]
Looks good in the first place for me.

Could you check if i forgot anything:

https://github.com/d-oit/hinode/tree/Language-1024

New params:

    enableLanguageSelectionStorage = true 

navbar:
ul language with id="language-selector"

new js file:

assets/js/critical/languageSelector.js

**Comment by markdumay** [comment]
Thanks for the PR! I suggest to simplify it a little. 

 - We store the preferred language whenever the user clicks on a language preference using the drop down.
 - When the user navigates to an url in a different language, the script redirects to the preferred language
 - The existing Hinode scripts should handle the drop down entirely (setting the active state)

I've also applied some linting fixes to the script. Below my suggestion, feel free to comment. I may have overlooked a few cases, or misinterpreted your intentions.
```js
{{- if site.Params.main.enableLanguageSelectionStorage -}}

(() => {
  'use strict'

  // Function to get the selected language from localStorage
  function getLanguage () {
    return localStorage.getItem('selectedLanguage')
  }

  // Function to set the selected language in localStorage
  function setLanguage (language) {
    localStorage.setItem('selectedLanguage', language)
  }

  // Function to apply the selected language to the website
  function applyLanguage (language, href) {
    if (document.documentElement.lang !== language) {
      if (href) {
        if (window.location.pathname !== href) {
          window.location.href = href
        }
      } else {
        window.location.href = '/' + language + '/'
      }
    }
  }

  // Event listener for language selection
  document.addEventListener('DOMContentLoaded', () => {
    const languageItems = document.querySelectorAll(
      '#language-selector .dropdown-item'
    )

    if (languageItems.length > 0) {
      const storedLanguage = getLanguage()

      // Redirect if the stored language differs from the active language
      if ((storedLanguage) && (document.documentElement.lang !== storedLanguage)) {
        languageItems.forEach(item => {
          if (item.getAttribute('hreflang') === storedLanguage) {
            applyLanguage(storedLanguage, item.getAttribute('href'))
          }
        })
      }

      // Update the stored language when the user selects a new one
      languageItems.forEach(item => {
        item.addEventListener('click', () => {
          const selectedLanguage = item.getAttribute('hreflang')

          if (selectedLanguage) {
            setLanguage(selectedLanguage)
          }
        })
      })
    }
  })
})()

{{- end -}}
```

**Comment by d-oit** [comment]
Adjust the code after I realized the language dropdown URL is the whole URL. My old cms only ref to the lang not the whole url with the lang.

My bad: I did not clean it up again and missed the comment in github fork repo. 

My special case:
I have two languages on my website with different content. 

blog content EN doesn't exist
Link in any blog content EN to the DE content -> redirect to EN if localStorage = EN

my easy workaround is remove the localStorage key before the link.

**Comment by d-oit** [comment]
Other special case:

my 'projects' folder content is only in **en**

url = domain/projects

_Case_: User stored in **lang = de**, search engine or other link ref = domain/de

my current working workaround:

_index.md

---
author: d.o.it
title: Projects
---
{{< redirectToProject >}}


**redirectToProject.html**
<script>    
    localStorage.removeItem("selectedLanguage")
    window.location.href = '/projects'
</script>

-> but this it special. i already used 

window.location.href = '/projects' Save language selection before changing

**Comment by d-oit** [comment]
The changes are ready to test:
https://github.com/d-oit/hinode

Description and test - del the selected language: 
http://localhost:1313/nl/blog/remove-selected-language/

**Comment by d-oit** [comment]
@markdumay hint: github action shows an error 
https://github.com/d-oit/hinode/actions/runs/10321674983/job/28575128107

it works locally without errors:
npm run start:example

**Comment by markdumay** [comment]
Thanks for the revisions! A couple of remarks / questions:
 1. Is there a specific use case for the shortcode`remove-selected-language.html`? If not, I'd suggest to drop it to reduce clutter.
 2. The `removeSelectedLanguage()` function does not exist in `languageSelector.js` (it is defined in `assets/js/optional/languageSelectorRemover.js`)
 3. When navigating to the site root (e.g. `localhost:1313/`), the site is redirected twice when `enableLanguageSelectionStorage` is set.

**Comment by markdumay** [comment]
Regarding the last two observations, I suggest a couple of changes to avoid the additional redirect:

Move the `const storedLanguage = getLanguage()` statement on line 36 to line 32 in  **`languageSelector.js`**.

Replace `removeSelectedLanguage();` in **`languageSelector.js`** with the following snippet:

```js
      // Redirect to the localized homepage
      const defaultLang = '{{ site.LanguageCode | default site.Language.Lang }}'
      let language = storedLanguage ? storedLanguage : defaultLang
      window.location.href = '/' + language + '/'
```

*Note: The dropdown for the language selector does not exist in the site root (due to `alias.html`). The code assumes the stored language is valid, using the default site language as fallback.*

Replace **`layouts/alias.html`** with the following content:

```html
{{ $lang := site.LanguageCode | default site.Language.Lang }}
<!DOCTYPE html>
<html {{ with $lang }}lang="{{ . }}"{{ end }}>
  <head>
    <title>{{ .Permalink }}</title>
    <link rel="canonical" href="{{ .Permalink }}">
    <meta name="robots" content="noindex">
    <meta charset="utf-8">
    <noscript><meta http-equiv="refresh" content="0; url={{ .Permalink }}"></noscript>
    {{ if site.Params.main.enableLanguageSelectionStorage }} 
      {{ partial "footer/scripts.html" (dict "filename" "js/critical.bundle.js" "match" "js/critical/**.js" "header" true "page" page) }}
    {{ else }}
      <script src='{{ partial "utilities/GetStaticURL" (dict "url" "js/alias.js") }}'></script>
    {{ end }}
  </head>
</html>
```

Update the module initialization in **`layouts/partials/footer/scripts.html`** (lines 28-43):

```
{{- $modules := "" -}}
{{ if $core }}
    {{ $config := $page.Scratch.Get "modules" }}
    {{ if not $config }}
        {{ errorf "partial [footer/scripts.html] - Cannot initialize module configuration" }}
    {{ end }}

    {{- $modules = $config.core -}}
    {{- if reflect.IsSlice $config.localize -}}
        {{- range $index, $mod := $modules -}}            
            {{- if in $config.localize $mod}}
                {{- $localize = true -}}
            {{- end -}}
        {{- end -}}
    {{- end -}}
{{- end -}}
```

**Comment by d-oit** [comment]
> Thanks for the revisions! A couple of remarks / questions:
> 
> 1. Is there a specific use case for the shortcode`remove-selected-language.html`? If not, I'd suggest to drop it to reduce clutter.

On my website I have projects pages only on the english website. And link to a page - only exist in eng

https://d-oit.github.io/blog/embed-google-calendar-responsive/

I didn't want to extend the link shortcode here to keep the whole thing optional. If it's too special, I could remove it for the theme.

> 2. The `removeSelectedLanguage()` function does not exist in `languageSelector.js` (it is defined in `assets/js/optional/languageSelectorRemover.js`)

Yes, it's a special case. I only need to load the code in special case - like to en/projects for example. 

> 3. When navigating to the site root (e.g. `localhost:1313/`), the site is redirected twice when `enableLanguageSelectionStorage` is set.

I tried to change/extend the core of the theme as little as possible. And didn't have that much experience with hugo / hinode either. 

 I'll take a look at the proposed changes.

**Comment by markdumay** [comment]
Ok, thanks for the clarifications! I think `remove-selected-language.html` is indeed an edge case - I suggest to remove it from the main theme (and keep it within your own custom/derived theme).

One more thing comes to mind, does the script still work when a specific page does not have a translation?

**Comment by d-oit** [comment]
> Ok, thanks for the clarifications! I think `remove-selected-language.html` is indeed an edge case - I suggest to remove it from the main theme (and keep it within your own custom/derived theme).

Check in without it - Works as described 👍🏻 
change ->  {{- partial "footer/scripts.html
 
**-** = string inserted, otherwise there was an error message

> One more thing comes to mind, does the script still work when a specific page does not have a translation?
localStorage key = nl
http://localhost:1313/fr/blog/composents/ -> http://localhost:1313/fr/blog/composents -> click on any other link -> nl page

http://localhost:1313/en/blog/ -> http://localhost:1313/nl/blog/

http://localhost:1313/nl/blog/components/ -> Page Not Found


Als Möglichkeit würde mir dazu einfallen

**Comment by markdumay** [comment]
Cool - ready to submit a PR to the base repository?

> One possibility that comes to mind is to set the hreflang link in the html -> special business case for search engines

The sitemap should already take care of this, its generated by Hugo out of the box.

**Comment by d-oit** [comment]
>>Cool - ready to submit a PR to the base repository?
Yes. But it's more your code then my 😄 

>> The sitemap should already take care of this, its generated by Hugo out of the box.

not quite - sitemap.xml is one file per language. But I can customize that myself if needed. 
https://translatepress.com/hreflang-tags-ultimate-guide/

**Comment by d-oit** [comment]
PR submitted. You could change any code if something wrong and close the issue.

**Comment by markdumay** [comment]
> not quite - sitemap.xml is one file per language. But I can customize that myself if needed.

Hugo does include the translations in each sitemap. Take a look at https://demo.gethinode.com/sitemap.xml for example, which is an index of the available sitemaps - one for each supported language. The following snippet shows the available translations for https://demo.gethinode.com/en/tags/blog/ as defined in https://demo.gethinode.com/en/sitemap.xml:

```xml
<url>
  <loc>https://demo.gethinode.com/en/tags/blog/</loc>
  <lastmod>2024-07-29T08:32:37+02:00</lastmod>
  <xhtml:link rel="alternate" hreflang="fr" href="https://demo.gethinode.com/fr/tags/blog/"/>
  <xhtml:link rel="alternate" hreflang="nl" href="https://demo.gethinode.com/nl/tags/blog/"/>
  <xhtml:link rel="alternate" hreflang="en" href="https://demo.gethinode.com/en/tags/blog/"/>
</url>
```

This will hopefully safe you some time. :-)

Note: not all browsers parse the sitemap XML correctly. I used the following command (zsh/bash) to retrieve the sitemap:

```bash
curl https://demo.gethinode.com/en/sitemap.xml > sitemap.xml
```

**Comment by markdumay** [comment]
> PR submitted. You could change any code if something wrong and close the issue.

Thanks again @d-oit! The PR has been merged and has been included in [release v0.25.5](https://github.com/gethinode/hinode/releases/tag/v0.25.5).
